kubernetes/structures_pod.go

```diff
@@ -663,8 +663,9 @@ func expandTolerations(tolerations []interface{}) []*v1.Toleration {
 		if value, ok := m["operator"]; ok {
 			ts[i].Operator = v1.TolerationOperator(value.(string))
 		}
-		if value, ok := m["toleration_seconds"]; ok {
-			ts[i].TolerationSeconds = ptrToInt64(int64(value.(int)))
+		sec := m["toleration_seconds"].(int)
+		if sec > 0 {
```

Comment:
The problem is that 0 and negative numbers are actually valid values; it means to evict immediately. If you treat 0 as missing which this seems to do, that means something different; missing means to never evict if it has already been scheduled but is no longer compatible with the toleration.

How do we distinguish between a user explicitly setting "toleration_seconds=0" vs omitting it? Unfortunately, thanks to [hashicorp/terraform#16796](https://github.com/hashicorp/terraform/issues/16796) we can't tell the difference.

Reply:
Right - I didn't really fully clue into that. 

I think we can use a hacky work around and put -1 for any immediate evictions, and use 0 as the unset value.

Reply:
2 mild problems with this proposal.

1) The [Kubernetes reference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#toleration-v1-core) for tolerationSeconds states "Zero **and negative values** will be treated as 0 (evict immediately) by the system." So even negative values like -1 actually have meaning.
2) Even if we intentionally deviated from the Kubernetes specification and said "negative numbers are disallowed/treated as null", the negative number would show up in the Terraform plan that is printed onscreen even when the user omits the tolerationSeconds value. If that user knows the above semantic meaning of negative numbers, they may get confused.

I actually had created a commit for this at https://github.com/chanzuckerberg/terraform-provider-kubernetes/commit/567a26787dc629c2b4997265dfa112892ce6f38f but haven't created a formal PR because I wasn't comfortable being the one to say "Terraform provider will intentionally deviate from the Kubernetes spec."

Reply:
Yeah -- this sucks, but i think it's a little different than that, if negative treat it as kubernetes would (which is thankfully identical to 0), if zero don't allow it, and if positive continue along...

So far local tests for this seem to work

Reply:
@mbarrien do you have some example tf i can test, so far i can't get this to fail?

I created a similar daemon set to your comment:

```
resource "kubernetes_daemonset" "test5" {
  metadata {
    name = "negative-seconds"

    labels {
      Test = "TfAcceptanceTest"
    }
  }

  spec {
    selector {
      match_labels {
        Test = "TfAcceptanceTest"
      }
    }

    template {
      metadata {
        labels {
          Test = "TfAcceptanceTest"
        }
      }

      spec {
        toleration {
          effect   = "NoSchedule"
          key      = "dedicated"
          operator = "Equal"
          value    = "kiam_master"
        }
        container {
          image = "redis:5.0.2"
          name  = "containername"
        }
      }
    }
  }
}
```

and got this yaml:

```
➭ kubectl get daemonsets -o yaml | grep toleration -A4
        tolerations:
        - effect: NoSchedule
          key: dedicated
          operator: Equal
          value: kiam_master
```

Reply:
My suggestion for this issue would be this:

* make `toleration_seconds` a TypeString.
* use a distinctive string value like `unspecified` to distinguish between the 0 value and not-specified. Handle this special value in the flatten / expander.
* Convert to / from integer in the flatten / expand functions using `strconv.ParseInt` / `strconv.FormatInt`
* in the validation function, enforce the Kubernetes specified range of valid values as well as the placeholder value `unspecified`. 

This solution has precedent and has been used to work around this exact same issue in the past.

Reply:
@alexsomesan I think this is addressed, not totally sure i got you right - now you must specify "unspecified" or a valid value; maybe i miss interpreted you though. 

This also unfortunatly only errors out during apply, ideally this would fail during plan:

```
Error: Error applying plan:

1 error(s) occurred:

* kubernetes_daemonset.test5: 1 error(s) occurred:

* kubernetes_daemonset.test5: invalid toleration_seconds must be int or "unspecified", got "":


```

Reply:
There's always making blank string "" be the unspecified value. It does have the property of being distinguishable from 0.